modules/imgproc/src/contours_approx.cpp

```diff
@@ -0,0 +1,353 @@
+{
+    const int len = (int)ares.size();
+    int i1;
+    // remove all previous items from the beginning
+    for( i1 = 1; i1 < len && ares[i1].s != 0; i1++ )
+    {
+        ares[i1 - 1].s = 0;
+    }
+    if( i1 == len )
+    {
+        // all points survived - skip to the end
+        return -1;
+    }
+    else
+    {
+        i1--;
+
+        int i2;
+        // remove all following items from the end
+        for( i2 = len - 2; i2 > 0 && ares[i2].s != 0; i2-- )
+        {
+            clear_until(ares, i2, len);
+            ares[i2 + 1].s = 0;
+        }
+        i2++;
+
+        // only two points left
+        if( i1 == 0 && i2 == len - 1 )
+        {
+            // find first non-removed element from the start
+            i1 = get_next_idx(ares, 0);
+            // append first item to the end
+            ares.push_back(ares[0]);
+            (ares.end() - 1)->removed = false;
+        }
+        return i1;
+    }
+}
+
+static void pass_cleanup(vector<ApproxItem> & ares, int start_idx)
+{
+    int count = 1;
+
+    const int len = (int)ares.size();
+    int first = start_idx;
+    for (int i = start_idx, prev = i; i < len; ++i)
+    {
+        ApproxItem & item = ares[i];
+        if (item.removed)
+            continue;
+        int next_idx = get_next_idx(ares, i);
+        if (next_idx == len || next_idx - i != 1)
+        {
+            if (count >= 2)
+            {
+                if( count == 2 )
+                {
+                    const int s1 = ares[prev].s;
+                    const int s2 = ares[i].s;
+
+                    if( s1 > s2 || (s1 == s2 && ares[prev].k <= ares[i].k) )
+                        /* remove second */
+                        clear_until(ares, get_next_idx(ares, prev), get_next_idx(ares, i));
+                    else
+                        /* remove first */
+                        clear_until(ares, first, i);
+                }
+                else
+                {
+                    first = get_next_idx(ares, first);
+                    clear_until(ares, first, i);
+                }
+            }
+            clear_until(ares, first, i);
+            first = i;
+            count = 1;
+        }
+        else
+        {
+            ++count;
+        }
+        prev = i;
+    }
+}
+
+} // <anonymous>::
+
+
+
+vector<Point> cv::approximateChainTC89(vector<schar> chain, const Point &origin, const int method)
+{
+    if (chain.size() == 0)
+    {
+        return vector<Point>({ origin });
+    }
+
+    const bool isApprox = method == CHAIN_APPROX_TC89_L1 || method == CHAIN_APPROX_TC89_KCOS;
+
+    ApproxItem root;
+    vector<ApproxItem> ares = pass_0(chain, origin, isApprox, method == CHAIN_APPROX_NONE);
+
+    if (isApprox)
+    {
+        CV_DbgAssert(ares.size() < (size_t)numeric_limits<int>::max());
+
+        // Pass 1.
+        // Determines support region for all the remained points */
+        for (int i = 0; i < (int)ares.size(); ++i)
```

Comment:
looks like if you replace indexes from int to `size_t` then all casts to int can be dropped. It's related to `pass_cleanup`, `clear_until`, `calc_dominance`, etc.

Reply:
Originally I tried to use `size_t`, but in some places arithmetics requires switching to signed type or comparing with signed values and VS doesn't like it. Also, perhaps 32-bit might be slightly faster or consume less memory? I've added debug assertion to check that main array size is less than max int for safety.

Reply:
Made these changes too.